reports-scheduler/src/main/java/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/handler/ReportsJobActionHandler.java

```diff
@@ -0,0 +1,192 @@
+              lock -> {
+                if (lock == null) {
+                  findFirstAvailableJob(searchHitsQueue);
+                } else {
+                  log.info("send job data(report_definition_id) to reporting core for execution");
+
+                  RestResponse restResponse =
+                      new BytesRestResponse(
+                          RestStatus.OK, hit.toXContent(JsonXContent.contentBuilder(), null));
+                  channel.sendResponse(restResponse);
+                }
+              },
+              exception -> {
+                channel.sendResponse(
+                    new BytesRestResponse(
+                        RestStatus.INTERNAL_SERVER_ERROR, "Failed to acquire lock"));
+                throw new IllegalStateException("Failed to acquire lock.");
+              }));
+    } else {
+      log.info("No jobs to execute");
+      channel.sendResponse(new BytesRestResponse(RestStatus.NO_CONTENT, "No jobs to execute"));
+    }
+  }
+
+  /**
+   * Once the report core(Kibana side) is done executing job, this function will be called to
+   * delete/release the lock, remove the job from job queue(ES index) TODO: update the .report index
+   *
+   * @param jobId the id of the scheduled job
+   */
+  public void updateJob(String jobId) {
+    // the lockId format is required by lockService to avoid conflict
+    final String lockId = JOB_QUEUE_INDEX_NAME + "-" + jobId;
+
+    // remove job from queue
+    final DeleteRequest deleteRequest = new DeleteRequest().index(JOB_QUEUE_INDEX_NAME).id(jobId);
+
+    client.delete(
+        deleteRequest,
+        new ActionListener<DeleteResponse>() {
+          @Override
+          public void onResponse(DeleteResponse deleteResponse) {
+            if (deleteResponse.getResult() == DocWriteResponse.Result.NOT_FOUND) {
+              channel.sendResponse(
+                  new BytesRestResponse(
+                      RestStatus.NOT_FOUND,
+                      String.format(Locale.ROOT, "Job id %s doesn't exist", jobId)));
+            } else if (deleteResponse.getResult() == DocWriteResponse.Result.DELETED) {
+              // delete lock of that job
+              lockService.deleteLock(
+                  lockId,
+                  ActionListener.wrap(
+                      deleted -> {
+                        log.debug("Deleted lock: {}", deleted);
+                        channel.sendResponse(
+                            new BytesRestResponse(
+                                RestStatus.OK,
+                                String.format(
+                                    Locale.ROOT,
+                                    "Job deleted from jobs queue. Job id: %s",
+                                    jobId)));
+                      },
+                      exception -> log.debug("Failed to delete lock", exception)));
+            }
+          }
+
+          @Override
+          public void onFailure(Exception e) {
+            channel.sendResponse(
```

Comment:
why is this always internal server error?

Reply:
They are all failure response from operations on database(delete or create). What do you suggest? (for the same comments you made else where)

Reply:
more appropriate HTTP error codes depending on the root cause of the issue? may be identify the root cause depending on "e"? say the lock is already released/deleted and second call fails then return "Resource not found".

Reply:
The scenario you mentioned is handled here. https://github.com/opendistro-for-elasticsearch/kibana-reports/blob/ff733205fe41408133bb6b3e85e5752bb5a50d95/reports-scheduler/src/main/java/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/handler/ReportsJobActionHandler.java#L164

Reply:
in what all scenarios onFailure() is expected?

Reply:
found some reference here about the Elasticsearch high-level-client exceptions, which is exactly what we are using here.
issue: https://github.com/elastic/elasticsearch/issues/30334
javaDoc: https://github.com/elastic/elasticsearch/blob/f56ab039aec803c22621b82bd4ea30441dd19b8c/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java#L240
> an {@link IOException} is usually thrown in case of failing to parse the REST response in the high-level REST client, the request times out or similar cases where there is no response coming back from the Elasticsearch server
an {@link ElasticsearchException} is usually thrown in case where the server returns a 4xx or 5xx error code. The high-level client then tries to parse the response body error details into a generic ElasticsearchException and suppresses the original {@link ResponseException}

Two types of exceptions are thrown by the `client`. My idea is to assign status code 502 to IOException and 503 to ElasticsearchExeception. What do you think?

> 502 Bad Gateway
The server was acting as a gateway or proxy and received an invalid response from the upstream server.[65]
503 Service Unavailable
The server cannot handle the request (because it is overloaded or down for maintenance). Generally, this is a temporary state.[66]
504 Gateway Timeout
The server was acting as a gateway or proxy and did not receive a timely response from the upstream server.[67]

Reply:
looks good